ash/src/extensions/khr/external_semaphore_fd.rs

```diff
@@ -0,0 +1,54 @@
+use crate::prelude::*;
+use crate::version::{DeviceV1_0, InstanceV1_0};
+use crate::vk;
+use std::ffi::CStr;
+use std::mem;
+
+#[derive(Clone)]
+pub struct ExternalSemaphoreFd {
+    handle: vk::Device,
+    external_semaphore_fd_fn: vk::KhrExternalSemaphoreFdFn,
+}
+
+impl ExternalSemaphoreFd {
+    pub fn new<I: InstanceV1_0, D: DeviceV1_0>(instance: &I, device: &D) -> Self {
+        let external_semaphore_fd_fn = vk::KhrExternalSemaphoreFdFn::load(|name| unsafe {
+            mem::transmute(instance.get_device_proc_addr(device.handle(), name.as_ptr()))
+        });
+        Self {
+            handle: device.handle(),
+            external_semaphore_fd_fn,
+        }
+    }
+
+    pub fn name() -> &'static CStr {
+        vk::KhrExternalSemaphoreFdFn::name()
+    }
+
+    #[doc = "<https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkImportSemaphoreFdKHR.html>"]
+    pub unsafe fn import_semaphore_fd(
+        &self,
+        import_info: &vk::ImportSemaphoreFdInfoKHR,
+    ) -> VkResult<()> {
+        self.external_semaphore_fd_fn
+            .import_semaphore_fd_khr(self.handle, import_info)
+            .result()
+    }
+
+    #[doc = "<https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkGetSemaphoreFdKHR.html>"]
+    pub unsafe fn get_semaphore_fd(&self, get_info: &vk::SemaphoreGetFdInfoKHR) -> VkResult<i32> {
+        let mut fd = -1;
```

Comment:
Is there a reason this is -1?

Reply:
Because `-1` is an invalid FD, `0` is not. It should anyway be extremely unlikely (if not against the spec) to return `VK_SUCCESS` without changing this value. Otherwise we should add an extra check that returns `Err` when the FD is invalid?

See also its `external_memory_fd` counterpart:

https://github.com/MaikKlein/ash/blob/d8d742348e7580e74a8385a2c0ff34a4b9db60e7/ash/src/extensions/khr/external_memory_fd.rs#L30

Reply:
Ah its a posix fd, okay then this should be fine as is. I don't think we should go out of our ways guard against incorrect drivers.

Reply:
Yup:https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_external_semaphore_fd.html

> An application using external memory may wish to synchronize access to that memory using semaphores. This extension enables an application to export semaphore payload to and import semaphore payload from **POSIX file descriptors**.